client/src/objects/drawable.rs

```diff
@@ -79,55 +78,60 @@ impl<'lua> Drawable<'lua> {
     /// Sets the geometry, and allocates a new surface.
     pub fn set_geometry(&mut self, lua: rlua::Context<'lua>, geometry: Area) -> rlua::Result<()> {
         use rlua::Error::RuntimeError;
-        let obj_clone = self.clone();
         let mut drawable = self.state_mut()?;
         let size_changed = drawable.geo != geometry;
         drawable.geo = geometry;
         if size_changed {
-            drawable.refreshed = false;
             drawable.surface = None;
-            drawable.wayland_shell = Some(
-                wayland_obj::create_xdg_toplevel(None)
-                    .expect("Could not construct an xdg toplevel for a drawable")
+            drawable.refreshed = false;
+        }
+        let size: Size = geometry.size;
+
+        if size_changed && size.width > 0 && size.height > 0 {
+            drawable.surface = Some(
+                ImageSurface::create(Format::ARgb32, size.width as i32, size.height as i32)
+                    .map_err(|err| RuntimeError(format!("Could not allocate {:?}", err)))?
             );
-            let size: Size = geometry.size;
-
-            if size.width > 0 && size.height > 0 {
-                let temp_file = tempfile::tempfile().expect("Could not make new temp file");
-                temp_file
-                    .set_len(size.width as u64 * size.height as u64 * 4)
-                    .expect("Could not set file length");
-                drawable.surface = Some(
-                    ImageSurface::create(Format::ARgb32, size.width as i32, size.height as i32)
-                        .map_err(|err| RuntimeError(format!("Could not allocate {:?}", err)))?
-                );
-                {
-                    let shell = drawable.wayland_shell.as_mut().unwrap();
-                    shell.set_size(size);
-                    shell
-                        .set_surface(&temp_file, size)
-                        .map_err(|_| RuntimeError(format!("Could not set surface for drawable")))?;
-                }
-                drawable.temp_file = temp_file;
-                Object::emit_signal(lua, &obj_clone, "property::surface".into(), Value::Nil)?;
-            }
+
+            drawable.buffer = Some(create_buffer(size).expect("Could not create buffer"));
         }
+        drop(drawable);
+        // TODO(ried): conditionally emit signals only for changed properties
```

Comment:
Can you do that in this patch set? It should be a pretty simple comparison logic flow.